src/asm/lexer.c

```diff
@@ -717,21 +717,94 @@ static void freeExpansion(struct Expansion *expansion)
 
 static bool isMacroChar(char c)
 {
-	return c == '@' || c == '#' || (c >= '0' && c <= '9');
+	return c == '@' || c == '#' || c == '<' || (c >= '0' && c <= '9');
+}
+
+/* forward declarations for readBracketedMacroArgNum */
+static int peek(void);
+static void shiftChar(void);
+static uint32_t readNumber(int radix, uint32_t baseValue);
+static bool startsIdentifier(int c);
+static bool continuesIdentifier(int c);
+
+static uint32_t readBracketedMacroArgNum(void)
+{
+	dbgPrint("Reading bracketed macro arg\n");
+	bool disableMacroArgs = lexerState->disableMacroArgs;
+	bool disableInterpolation = lexerState->disableInterpolation;
+
+	lexerState->disableMacroArgs = false;
+	lexerState->disableInterpolation = false;
+
+	uint32_t num = 0;
+	int c = peek();
+	bool empty = false;
+
+	if (c >= '0' && c <= '9') {
+		num = readNumber(10, 0);
+	} else if (startsIdentifier(c)) {
+		char symName[MAXSYMLEN + 1];
+		size_t i = 0;
+
+		for (; continuesIdentifier(c); c = peek()) {
+			if (i < sizeof(symName))
+				symName[i++] = c;
+			shiftChar();
+		}
+
+		if (i == sizeof(symName)) {
+			warning(WARNING_LONG_STR, "Symbol name too long\n");
+			i--;
+		}
+		symName[i] = '\0';
+
+		struct Symbol const *sym = sym_FindScopedSymbol(symName);
+
+		if (!sym)
+			fatalerror("Bracketed symbol \"%s\" does not exist\n", symName);
+		else if (!sym_IsNumeric(sym))
+			fatalerror("Bracketed symbol \"%s\" is not numeric\n", symName);
+
+		num = sym_GetConstantSymValue(sym);
+	} else {
+		empty = true;
+	}
+
+	c = peek();
+	if (c != '>')
+		fatalerror("Invalid character in bracketed macro argument %s\n", printChar(c));
```

Comment:
Do these need to be fatal? An invalid char could cause the bracketing to implicitly stop, and an empty/0 argument to be expanded to nothing.

Reply:
I made them fatal to match the fatal "`Invalid macro argument '\0'`" and "`Macro argument '\%c' not defined`" errors. IMO either they should all be fatal, or all become empty expansions.

Reply:
I'd prefer the latter, though it can be a separate issue then.

Reply:
Non-fatal errors for them all would match interpolations, where `{undefinedsymbol}` and `{SomeLabel}` are merely `error`s. I can do that as a separate PR/commit.

Reply:
See #850 :P